### PR TITLE
use-cases: Add token account

### PR DESCRIPTION
**Pull request by j-mueller**
* The token account contract lets you pay "to a token", by releasing the
  funds locked with it to whoever presents the token.

### CODE REVIEW

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+-- | Pay some money to the given token
+pay :: (AsContractError e, HasWriteTx s) => AccountOwner -> Value -> Contract s e TxId
+pay owner vl =
+    let ds = Ledger.DataScript (PlutusTx.toData owner)
+    in writeTxSuccess (payToScript vl address ds)
+
+ownsTxOut :: AccountOwner -> TxOut -> Bool
+ownsTxOut owner txout =
+    Just owner == (Ledger.txOutData txout >>= PlutusTx.fromData . Ledger.getDataScript)
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: I'd bracket this a bit more - I don't think many people know what the relative precedence of `.` and `>>=` is...

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    :: ( HasUtxoAt s )
+    => AccountOwner
+    -> Contract s e Value
+balance owner = do
+    utxos <- utxoAt (Scripts.scriptAddress scriptInstance)
+    let inner =
+            foldMap (view Ledger.outValue)
+            $ filter (ownsTxOut owner)
+            $ fmap snd
+            $ Map.toList
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: Nit: you can use `Map.filter` and then `foldMap` will work on the map values so you can drop the `Map.toList` and `fmap snd`.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+-- | Create a new token and return its 'AccountOwner' information.
+newAccount
+    :: ( HasWatchAddress s
+       , HasWriteTx s
+       , AsContractError e
+       )
+    => PubKey
+    -> Contract s e AccountOwner
+newAccount pk = do
+    cur <- Currency.forgeContract pk [("token-account", 1)]
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: Hm, surely this shouldn't be a constant string?

File: plutus-use-cases/test/Spec/TokenAccount.hs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        (assertNoFailedTransactions
+        /\ assertDone w1 (const True) "wallet 1 should be done"
+        /\ walletFundsChange w1 (Ada.lovelaceValueOf (-10) <> theToken))
+        (handleBlockchainEvents w1)
+
+    , checkPredicate "Redeem all funds"
+        (initialiseAccount >>= Accounts.redeem (walletPubKey w1))
+        (assertNoFailedTransactions
+        /\ assertDone w1 (const True) "wallet 1 should be done"
+        /\ walletFundsChange w1 theToken)
```
Comment (michaelpj) [inline plutus-use-cases/test/Spec/TokenAccount.hs]: I was a bit confused by no ada check here, but then I realised you're claiming the money back to the original owner. Maybe have a test where you send the token to someone else and they redeem?

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+redeemTx
+    :: ( HasUtxoAt s )
+    => AccountOwner
+    -> PubKey
+    -> Contract s e UnbalancedTx
+redeemTx owner pk = do
+    utxos <- utxoAt (Scripts.scriptAddress scriptInstance)
+    let filterByOwner _ = ownsTxOut owner
+        tx = TypedTx.collectFromScriptFilter filterByOwner utxos scriptInstance ()
+    -- TODO. Replace 'PubKey' with a more general 'Address' type of output?
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: Not sure I understand what you want generalize? Is it that you want to be able to send it to a script?

Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: At the moment the two constraints "token X must be present" and "amount X must be paid to Y" are indistinguishable in `UnbalancedTx` and the comment was meant to suggest that there should be a way of saying that a token (value) must be moved by the transaction, without saying anything about the outputs it comes from or goes to.

Right now "token X must be present" is implemented as a pubkey output with a value of exactly X, but this means that (a) we need to know which pubkey to use for the output in `redeemTx`, which isn't a concern of the contract itself, and (b) yes we can't send it to a script. Although if we had the "token X must be present" constraint then (b) wouldn't matter anymore.

I would really like to be able to write these contracts without any reference to public key or script addresses (as far as ownership is concerned), just tokens.

File: plutus-use-cases/test/Spec/TokenAccount.hs (hunk 4 above)
Comment (j-mueller) [reply plutus-use-cases/test/Spec/TokenAccount.hs]: Done

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+scriptInstance :: Scripts.ScriptInstance TokenAccount
+scriptInstance =
+    Scripts.Validator @TokenAccount $$(PlutusTx.compile [|| validate ||]) $$(PlutusTx.compile [|| wrap ||])
+    where
+    wrap = Scripts.wrapValidator @AccountOwner @()
+
+address :: Address
+address = Scripts.scriptAddress scriptInstance
+
+-- | Pay some money to the given token
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]:
```suggestion
-- | Pay some money to the given token account
```

Comment (michaelpj) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: This is slightly confusing given that the type is `AccountOwner` not `Account`, but it's clear enough.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import           Ledger.TxId                                       (TxId)
+import           Ledger.Typed.Scripts                              (ScriptType (..))
+import qualified Ledger.Typed.Scripts                              as Scripts
+import qualified Ledger.Validation                                 as V
+import           Ledger.Value                                      (CurrencySymbol, TokenName, Value)
+import qualified Ledger.Value                                      as Value
+
+import qualified Language.PlutusTx.Coordination.Contracts.Currency as Currency
+
+newtype AccountOwner = AccountOwner { unAccountOwner :: (CurrencySymbol, TokenName) }
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs]: Yeah that would make sense!

**Pull request merged**